# Knowledge and Data: Practical Assignment 1 
## Manipulate Propositional Logic and Simple Knowledge Graph Logic

- YOUR NAME: Matúš Hába

- YOUR VUNetID: mha473

*(If you do not provide your NAME and VUNetID we will not accept your submission.)*

### Learning objectives

At the end of this exercise you should be able to perform some simple manipulations of two different formal systems: 
1. You will be able to transform Propositional Logic statements from one syntactic form to semantically equivalent  alternative representation.
2. You will be able to evaluate a Propositional Logic statements given some validation of the propositional variables. 
3. You will be able to transform one Knowledge Graph into an equivalent one with a different Syntax 
4. You will be able to calculate simple entailment between two Knowledge Graph

### Practicalities

Follow this Notebook step-by-step. Some of the exercises are already 
fully filled in to give you some examples of how to do things. Others are
partially filled, and require you to fill in some gaps. And others, you 
will have to fully program yourself. 

Of course, you can do the exercises in any Programming Editor of your liking. 
But you do not have to. Feel free to simply write code in the Notebook. When 
everythink is filled in and works, save the Notebook and submit it 
as a Jupyter Notebook (i.e. with an ipynb extension). **Please use as name of the 
Notebook your studentID+Assignment1.ipynb**.  

We will not evaluate the programming style of your solutions. Yet we do look whether your solutions suggests an understanding, and whether they yield the correct output. 

Note that all notebooks will automatically be checked for plagiarism: while similar answers can be expected, it is not allowed to directly copy the solutions from fellow students or TAs, or from the examples discussed during the lectures. Similarly, sharing your solutions with your peers is not allowed.

## Exercises related to Propositional Logic 

We will use a very simple, though crude and ugly, representation for Propositional Logic: Propositions are strings, and formulas are lists, starting with an operator as first element in the list, and the subformulas as second and (if appropriate) third argument. (Note that this is slightly different than the definition in the lecture, where Propositions are lists with one argument). These are formulas in Prefix Notation.

In the following part, we will read several such formulas from a file (`PLknowlegebase.txt`).

For simplicity, we will only consider formulas with any of the three variables P, Q, and R, and binary operators ">>" (implies), "&" (and), "|" (or), and "~" (not). 

Here are some formulas as examples.

(~R & P) >> Q

~(((P | ~Q) >> R) >> (P & R)) 

~(P >> Q) | (R >> P) 

(P >> Q) | (Q >> ~P)


The first thing you will have to do is to open the file. The code below will read every line of the file into a list called *content*.

Do *not* forget to run the following cell, otherwise, the file will not be opened. import os 


In [78]:
import os 
from logic import *

In [79]:
fileDir = os.path.dirname(os.path.realpath('__file__'))
filename = os.path.join(fileDir, 'PLknowledgebase.txt')
print(filename)
with open(filename) as file:
    content = file.readlines()

print(content)

C:\Users\matus\Documents\VU\Knowledge and Data\knowledge-data-vu\Assignments\Assignment_1\PLknowledgebase.txt
['(~R & P) >> Q\n', '~(((P | ~Q) >> R) >> (P & R)) \n', '~(P >> Q) | (R >> P) \n', '(P >> Q) | (Q >> ~P)']


In [80]:
P, Q, R = vars('P', 'Q', 'R')

for line in content:
    formula = eval(line.strip())
    print()
    print(formula)
    print(formula.is_tautology())
    print(type(formula))
    print(formula.children[0])


(¬R ^ P) => Q
False
<class 'logic.Implies'>
¬R ^ P

¬(((P v ¬Q) => R) => (P ^ R))
False
<class 'logic.Not'>
((P v ¬Q) => R) => (P ^ R)

¬(P => Q) v (R => P)
False
<class 'logic.Or'>
¬(P => Q)

(P => Q) v (Q => ¬P)
True
<class 'logic.Or'>
P => Q


The last procedure makes use of the inbuilt properties of the logic package. Let us do similar stuff ourselves. 

### Task 1: (10 Points) Complete the procedure *prefix* that transforms the formula into Prefix notation, i.e., the operator "sits" in front of the two arguments.

Use different symbols for the prefix operators: IMP for implication, AND for conjunction, OR for disjunction and NEG for negation. Look at the previous examples if you are unsure how to continue.

In [81]:
def prefix(s):
    if type(s) is And:
        return "AND(" + prefix(s.children[0]) + "," + prefix(s.children[1]) + ")"
    
    elif type(s) is Or:
        return "OR(" + prefix(s.children[0]) + "," + prefix(s.children[1]) + ")"
    
    elif type(s) is Not:
        return "NEG(" + prefix(s.children[0]) + ")"
    
    elif type(s) is Implies:
        return "IMP(" + prefix(s.children[0]) + "," + prefix(s.children[1]) + ")"
    
    else:
        return str(s)

Execute the following line of code to check whether your procedure works correctly or not.

In [82]:
for line in content:
    formula = eval(line.strip())
    spacing = 30 - len(str(formula))
    print(str(formula) + " "*spacing + "== " + prefix(formula))

(¬R ^ P) => Q                 == IMP(AND(NEG(R),P),Q)
¬(((P v ¬Q) => R) => (P ^ R)) == NEG(IMP(IMP(OR(P,NEG(Q)),R),AND(P,R)))
¬(P => Q) v (R => P)          == OR(NEG(IMP(P,Q)),IMP(R,P))
(P => Q) v (Q => ¬P)          == OR(IMP(P,Q),IMP(Q,NEG(P)))


If you run your procedure on the input data, you should get something like the following output: 

    (¬R ^ P) => Q                 == IMP(AND(NEG(R),P),Q)
    ¬(((P v ¬Q) => R) => (P ^ R)) == NEG(IMP(IMP(OR(P,NEG(Q)),R),AND(P,R)))
    ¬(P => Q) v (R => P)          == OR(NEG(IMP(P,Q)),IMP(R,P))
    (P => Q) v (Q => ¬P)          == OR(IMP(P,Q),IMP(Q,NEG(P)))

### Task 2: (20 Points) Write a procedure *evaluate* that calculates the truth value of a formula. 

Given input formula _s_, and the three truth values for boolean variable P, Q, and R, write a procedure that tells us whether _s_ is true or false.

This is a procedural way to implement the "meaning" of the operators (which you can read from the Truth Tables)

In [83]:
def evaluate(s, p, q, r):
    if type(s) is Or:
        return evaluate(s.children[0],p,q,r) or evaluate(s.children[1],p,q,r)
    
    elif type(s) is And:
        return evaluate(s.children[0],p,q,r) and evaluate(s.children[1],p,q,r)
    
    elif type(s) is Not:
        return not evaluate(s.children[0],p,q,r)
        
    elif type(s) is Implies:
        return not evaluate(s.children[0],p,q,r) or evaluate(s.children[1],p,q,r)
    
    elif type(s) is Variable:       
        return {
            'P': p,
            'Q': q,
            'R': r,
        }[str(s)]
    
    print("Something went wrong") 

Execute the following line of code to check whether your procedure works correctly or not.

In [84]:
for l in content:
    formula = l.strip()
    spacing = 30 - len(str(formula))
    print(str(formula) + " "*spacing + str(evaluate(s=eval(formula), p=True, q=True, r=False)))

(~R & P) >> Q                 True
~(((P | ~Q) >> R) >> (P & R)) False
~(P >> Q) | (R >> P)          True
(P >> Q) | (Q >> ~P)          True


If you run your program on the given input, you should get an answer like this:

    (~R & P) >> Q                 True
    ~(((P | ~Q) >> R) >> (P & R)) False
    ~(P >> Q) | (R >> P)          True
    (P >> Q) | (Q >> ~P)          True

### Task 3: (10 Points) Write a procedure *tautology* that calculates whether a given formula (with maximally three variables P,Q and R) is a tautology 
Hint: Use the *evaluate* function you just wrote. Do *not* use the 'is_tautology' method seen earlier.

In [85]:
def tautology(s):
    values = [True, False]
    for p in values:
        for q in values:
            for r in values:
                if not evaluate(s,p,q,r):
                    return False
    return True

Execute the following line of code to check whether your procedure works correctly or not.

In [86]:
print("Formula" + 23*" " + "Tautology")
for l in content:
    formula = l.strip()
    spacing = 30 - len(str(formula))
    print(str(formula) + " "*spacing + (repr(tautology(eval(formula)))))

Formula                       Tautology
(~R & P) >> Q                 False
~(((P | ~Q) >> R) >> (P & R)) False
~(P >> Q) | (R >> P)          False
(P >> Q) | (Q >> ~P)          True


When you evaluate your code on the provided input, you should get something like this:

    Formula                       Tautology
    (~R & P) >> Q                 False
    ~(((P | ~Q) >> R) >> (P & R)) False
    ~(P >> Q) | (R >> P)          False
    (P >> Q) | (Q >> ~P)          True

## Exercises related to Simple Knowledge Graph Logic 

We will use a very simple, though crude and ugly, representation for Simple Knowledge Graph Logic: Resources are strings, triples are tuples of resources, and knowledge graphs are lists of tuples. So, a triple ( s p o ) will be represented as a tuple `(s, p, o)`, and a Knowledge Graph with two triples (s1, p1, o1) and (s2, p2, o2) as a list containing two tuples: `[(s1,p1,o1), (s2,p2,o2)]`.

Every line in the file `knowledgegraph.txt` contains a knowledge graph (one list). 

Run the following code to load the Knowledge Graphs in knowledgegraph.txt

In [87]:
import os 
fileDir = os.path.dirname(os.path.realpath('__file__'))
filename = os.path.join(fileDir, 'knowledgegraph.txt')
with open(filename) as file:
    content = [[t for t in eval(g)] for g in file.readlines()]

Maybe you want to write a function to print this graph for convenience ... (no extra points, though) 

The next two exercises are about syntactic transformation of Knowledge Graphs. The two target formats are simplified versions of syntaxes for the language RDF that we will introduce later. 

### Task 4a: (10 Points) Write a procedure to transform the Knowledge Graphs into N-triple syntax. 
N-triple is a simple (the most simple?) representation format for representing Knowledge Graphs. 

A triple (s,p,o) is simply written as as follows:

     <s> <p> <o> . 
     
As one string, with blanks between s,p and o, with < and > around the variables, and a dot after the triple. 

First, write a function that transforms a triple ('s','p','o') into a single string "\<s> \<p> \<o> ."

In [88]:
def as_ntriple(triple):
    return "<{}> <{}> <{}>.".format(*triple)

In [89]:
def as_ntriples(graph):
    return [as_ntriple(triple) for triple in graph]

Execute the following line of code to check whether your procedure works correctly or not.

In [90]:
for graph in content:
     print(as_ntriples(graph))

['<s> <p> <o>.', '<s> <p1> <o2>.', '<s2> <p2> <o>.']
['<Netherlands> <name> <"The Netherlands">.', '<Netherlands> <has_capital> <Amsterdam>.', '<Amsterdam> <type> <Capital>.', '<Amsterdam> <has_population> <"921.402">.']
['<Netherlands> <name> <"The Netherlands">.', '<Netherlands> <has_capital> <Amsterdam>.', '<Amsterdam> <type> <Capital>.', '<Amsterdam> <has_population> <"921.402">.', '<Capital> <type> <City>.', '<Netherlands> <neighbours> <Belgium>.', '<Netherlands> <type> <Country>.']


Your answer should look similar to this: <br />
['\<s> \<p> \<o> .', '\<s> \<p1> \<o2> .', '\<s2> \<p2> \<o> .']<br/>
['\<Netherlands> \<name> \<"Netherlands"> .', '\<Netherlands> \<has_capital> \<Amsterdam> .', '\<Amsterdam> \<type> \<Capital> .']<br/>
['\<Netherlands> \<name> \<"Netherlands"> .', '\<Netherlands> \<has_capital> \<Amsterdam> .', '\<Amsterdam> \<type> \<Capital> .', '\<Capital> \<type> \<City> .', '\<Netherlands> \<neighbours> \<Belgium> .', '\<Netherlands> \<type> \<Country> .']

### Task 4b (10 Points) Adjust your procedure to cope with literal objects

In reality, the N-triples format specifies that only objects must be embedded between angled brackets; literal values, such as strings and numbers, are to be represented as string values, for example "\<s> \<p> \"12\" .". Since knowledge graphs are construced with an object-centric language (RDF), we only have to consider the last element of a triple.

Redefine the function _as\_triple_ to make it output valid N-triples.

In [91]:
def as_ntriple(triple):
    return f"<{triple[0]}> <{triple[1]}> {triple[2] if triple[2].startswith('"') else f'<{triple[2]}>'}." 

Execute the following line of code to check whether your procedure works correctly or not.

In [92]:
for graph in content:
     print(as_ntriples(graph))

['<s> <p> <o>.', '<s> <p1> <o2>.', '<s2> <p2> <o>.']
['<Netherlands> <name> "The Netherlands".', '<Netherlands> <has_capital> <Amsterdam>.', '<Amsterdam> <type> <Capital>.', '<Amsterdam> <has_population> "921.402".']
['<Netherlands> <name> "The Netherlands".', '<Netherlands> <has_capital> <Amsterdam>.', '<Amsterdam> <type> <Capital>.', '<Amsterdam> <has_population> "921.402".', '<Capital> <type> <City>.', '<Netherlands> <neighbours> <Belgium>.', '<Netherlands> <type> <Country>.']


Your answer should look similar to this: <br />
['\<s> \<p> \<o> .', '\<s> \<p1> \<o2> .', '\<s2> \<p2> \<o> .']<br/>
['\<Netherlands> \<name> "Netherlands" .', '\<Netherlands> \<has_capital> \<Amsterdam> .', '\<Amsterdam> \<type> \<Capital> .',  '\<Amsterdam> \<has_population> "921.402" .']<br/>
['\<Netherlands> \<name> "Netherlands" .', '\<Netherlands> \<has_capital> \<Amsterdam> .', '\<Amsterdam> \<type> \<Capital> .', '\<Capital> \<type> \<City> .', '\<Netherlands> \<neighbours> \<Belgium> .', '\<Netherlands> \<type> \<Country> .',  '\<Amsterdam> \<has_population> "921.402" .']

### Task 5: (20 Points) Write a procedure to evaluate whether a Triple (s p o) is entailed by your Knowledge Graph

Remember that we can calculate whether a triple is entailed by a Knowledge Graph (w.r.t Simple Knowledge Graph Logic) by simply checking whether the triple is part of the knowledge graph.  

The triples, on which you can test your implementations are located in the file: *entailedTriples.txt*. You need to open the file and check for all triples whether it is entailed by the knowledge graphs in *knowledgegraph.txt* (which is already opened). Hint: you can use *eval* to convert a raw string to a list.

You might want to add auxiliary functions in extra cells if necessary.

In [93]:
def tripleEntailedBy(graph, triple, output=True):
    # NOTE: as the assignment description does not go into detail explaining where to print the expected output or whether to return anything,
    # I have decided that this function will be returning a boolean value and printing an adequate evaluation that is expected

    #NOTE: since the assignment description does not specify whether we can reuse functions, 
    # this function has been adapted to fit the needs of assignment 6 as well as this one
    
    if triple in graph:
        if output:
            print(graph, "entails", triple)
        return True
    if output:
        print(graph, "does NOT entail", triple)
    return False

Provide some code to check that your function works correctly.  

In [94]:
filename = os.path.join(fileDir, "entailedTriples.txt")
with open(filename) as file:
    triples = [eval(t) for t in file.readlines()]

for graph in content:
    for triple in triples:
        tripleEntailedBy(graph, triple)

[('s', 'p', 'o'), ('s', 'p1', 'o2'), ('s2', 'p2', 'o')] entails ('s2', 'p2', 'o')
[('s', 'p', 'o'), ('s', 'p1', 'o2'), ('s2', 'p2', 'o')] does NOT entail ('Netherlands', 'name', '"The Netherlands"')
[('Netherlands', 'name', '"The Netherlands"'), ('Netherlands', 'has_capital', 'Amsterdam'), ('Amsterdam', 'type', 'Capital'), ('Amsterdam', 'has_population', '"921.402"')] does NOT entail ('s2', 'p2', 'o')
[('Netherlands', 'name', '"The Netherlands"'), ('Netherlands', 'has_capital', 'Amsterdam'), ('Amsterdam', 'type', 'Capital'), ('Amsterdam', 'has_population', '"921.402"')] entails ('Netherlands', 'name', '"The Netherlands"')
[('Netherlands', 'name', '"The Netherlands"'), ('Netherlands', 'has_capital', 'Amsterdam'), ('Amsterdam', 'type', 'Capital'), ('Amsterdam', 'has_population', '"921.402"'), ('Capital', 'type', 'City'), ('Netherlands', 'neighbours', 'Belgium'), ('Netherlands', 'type', 'Country')] does NOT entail ('s2', 'p2', 'o')
[('Netherlands', 'name', '"The Netherlands"'), ('Netherla

The expected output is a list of relations between graphs and triples<br />
[('s','p','o'),('s','p1','o2'),('s2','p2','o')] entails ('s2','p2','o')<br />
[('s','p','o'),('s','p1','o2'),('s2','p2','o')] does NOT entail ('Netherlands','name','"The Netherlands"')<br />
and so on for all graphs and triples. 

### Task 6: (20 Points) Write a procedure to evaluate whether one Knowledge Graph is entailed by another Knowledge Graph

Remember that we can calculate whether a triple is entailed by another graph (w.r.t Simple Knowledge Graph Logic) by simply checking whether the former is a subgraph of the latter. The same holds for graphs.

Write a procedure to check for all of graphs in `entailedGraphs.txt` whether they are entailed by those in `knowledgegraph.txt` (which is already opened).

You might want to add auxiliary functions in extra cells if necessary.

In [95]:
def graphEntailedBy(graphEntails, graphEntailed):
    for triple in graphEntailed:
        # NOTE: the assignment does not specify whether we can reuse functions from previous tasks, 
        # therefore I saw no reason for redeclaring it in an auxiliary cell and chose to reuse (adapt) tripleEntailedBy() directly
        if not tripleEntailedBy(graphEntails, triple, False):
            print(graphEntails, "does NOT entail", graphEntailed)
            return False
    
    print(graphEntails, "entails", graphEntailed)
    return True

Provide some code to check that your function works correctly.  

In [96]:
filename = os.path.join(fileDir, 'entailedGraphs.txt')
with open(filename) as file:
    graphs = [[t for t in eval(g)] for g in file.readlines()]

for graphEntails in content:
    for graphEntailed in graphs:
        graphEntailedBy(graphEntails, graphEntailed)

[('s', 'p', 'o'), ('s', 'p1', 'o2'), ('s2', 'p2', 'o')] entails [('s2', 'p2', 'o'), ('s2', 'p2', 'o')]
[('s', 'p', 'o'), ('s', 'p1', 'o2'), ('s2', 'p2', 'o')] entails [('s', 'p', 'o'), ('s2', 'p2', 'o')]
[('s', 'p', 'o'), ('s', 'p1', 'o2'), ('s2', 'p2', 'o')] does NOT entail [('Netherlands', 'name', '"The Netherlands"'), ('Netherlands', 'has_capital', 'Amsterdam'), ('Capital', 'a', 'City'), ('Netherlands', 'neighbours', 'Belgium')]
[('s', 'p', 'o'), ('s', 'p1', 'o2'), ('s2', 'p2', 'o')] does NOT entail [('Netherlands', 'has_capital', 'Amsterdam'), ('Capital', 'a', 'Country'), ('Netherlands', 'neighbours', 'Belgium')]
[('Netherlands', 'name', '"The Netherlands"'), ('Netherlands', 'has_capital', 'Amsterdam'), ('Amsterdam', 'type', 'Capital'), ('Amsterdam', 'has_population', '"921.402"')] does NOT entail [('s2', 'p2', 'o'), ('s2', 'p2', 'o')]
[('Netherlands', 'name', '"The Netherlands"'), ('Netherlands', 'has_capital', 'Amsterdam'), ('Amsterdam', 'type', 'Capital'), ('Amsterdam', 'has_pop

The expected output is a list of relations between graphs and graphs<br />
[('s', 'p', 'o'), ('s', 'p1', 'o2'), ('s2', 'p2', 'o')] entails [('s2', 'p2', 'o'), ('s2', 'p2', 'o')]<br />
[('s', 'p', 'o'), ('s', 'p1', 'o2'), ('s2', 'p2', 'o')] entails [('s', 'p', 'o'), ('s2', 'p2', 'o')]<br />
[('s', 'p', 'o'), ('s', 'p1', 'o2'), ('s2', 'p2', 'o')] does NOT entail [('Netherlands', 'name', '"The Netherlands"'), ('Netherlands', 'has_capital', 'Amsterdam'), ('Capital', 'a', 'City'), ('Netherlands', 'neighbours', 'Belgium')]
<br />
and so on for all graphs and triples. 